### PR TITLE
Refactoring use of GameObject

### PR DESCRIPTION
**Pull request by DunderRoffe**
- Documentation for GameObject
- Handling all GameObject pointers as std::shared_ptr<GameObject>

### CODE REVIEW

File: includes/GameObject.h [hunk 1]
```diff
@@ -98,57 +100,152 @@ class GameObject : public IDrawable {
      */
     virtual void renderShadow(std::shared_ptr<ShaderProgram> &shaderProgram);
 
+    /**
+     * Add a IRenderComponent to the GameObject.
```
Comment (SiXoS) [inline includes/GameObject.h:104]: Unecessary comment, will just be outdated if we change type name

File: includes/GameObject.h [hunk 2]
```diff
@@ -98,57 +100,152 @@ class GameObject : public IDrawable {
...
+     */
     void addRenderComponent(IRenderComponent* renderer);
+
+    /**
+     * Add a component to the GameObject.
```
Comment (SiXoS) [inline includes/GameObject.h:109]: Unecessary comment

File: includes/GameObject.h [hunk 3]
```diff
@@ -98,57 +100,152 @@ class GameObject : public IDrawable {
...
      *
      * @param type The event to call on the components
      * @param data // TODO(Bubbad) Do all events need to involve another gameobject?
      */
-    void callEvent(EventType type, GameObject* data);
-
+    void callEvent(EventType type, std::shared_ptr<GameObject> &data);
 
+    /**
+     * Get the model matrix of the GameObject.
```
Comment (SiXoS) [inline includes/GameObject.h:129]: Unecessary comment

File: includes/GameObject.h [hunk 4]
```diff
@@ -98,57 +100,152 @@ class GameObject : public IDrawable {
...
+     */
     chag::float3 getRelativeScale();
+
+    /**
+     * Get scale of the GameObject relative to default scale.
+     */
     chag::float3 getAbsoluteScale();
+
+    /**
+     * Set the absoulte scale of the GameObject.
```
Comment (SiXoS) [inline includes/GameObject.h:159]: Isn't this relative to the parent?

File: includes/GameObject.h [hunk 5]
```diff
@@ -98,57 +100,152 @@ class GameObject : public IDrawable {
...
     void update(chag::float4x4 update_matrix);
 
+    /**
+     * Get the scale of the GameObject relative to its parent.
+     * If there is no parent, this is the same as getAbsoluteScale().
+     */
     chag::float3 getRelativeScale();
+
+    /**
+     * Get scale of the GameObject relative to default scale.
```
Comment (SiXoS) [inline includes/GameObject.h:154]: Original rotation might be better than default? Or model rotation or something

File: includes/GameObject.h [hunk 6]
```diff
@@ -98,57 +100,152 @@ class GameObject : public IDrawable {
...
     void setScale(chag::float3 s);
 
+    /**
+     * Get the rotation of the GameObject relative to its parent.
+     * If there is no parent, this is the same as getAbsoluteRotation().
+     */
     chag::Quaternion getRelativeRotation();
+
+    /**
+     * Get the rotation of the GameObject relative to default rotation.
```
Comment (SiXoS) [inline includes/GameObject.h:170]: original?

File: includes/GameObject.h [hunk 7]
```diff
@@ -98,57 +100,152 @@ class GameObject : public IDrawable {
...
+
+    /**
+     * Sets the absolute location of the GameObject.
+     */
     void setLocation(chag::float3 l);
 
-    void addChild(GameObject* child);
+    /**
+     * Tries to make a parent child relationship between this GameObject and a
+     * given child GameObject. Requires that the child being added dose not
```
Comment (SiXoS) [inline includes/GameObject.h:202]: does*

File: includes/GameObject.h [hunk 8]
```diff
@@ -157,26 +254,54 @@ class GameObject : public IDrawable {
...
     std::vector<Triangle*> getTriangles();
 
+    /**
+     * Get a AABB fitting the GameObjects Mesh.
+     * The AABB is transformed by the GameObjects model matrix.
+     */
     AABB getTransformedAABB();
+
+    /**
+     * Get the minimal boundng sphere that contains the Mesh of the GameObject.
```
Comment (SiXoS) [inline includes/GameObject.h:272]: bounding

File: includes/GameObject.h [hunk 9]
```diff
@@ -157,26 +254,54 @@ class GameObject : public IDrawable {
...
      * Marks the GameObject as dirty, indicating that
-     * it should be removed at some point.
+     * the GameObject will be removed soon.
+     *
+     * CAUTION: THIS ACTION IS IRREVERSIBLE
      */
     void makeDirty();
+
+    /**
+     * Check if the GameObject is dirty or not.
```
Comment (SiXoS) [inline includes/GameObject.h:303]: Unecessary comment

File: includes/GameObject.h [hunk 10]
```diff
@@ -157,26 +254,54 @@ class GameObject : public IDrawable {
...
+
     /**
      * NOTE: The triangles have not been transformed.
      *
      * @return A list of all the triangles in the GameObject
      */
     std::vector<Triangle*> getTriangles();
 
+    /**
+     * Get a AABB fitting the GameObjects Mesh.
```
Comment (SiXoS) [inline includes/GameObject.h:266]: an*

**Review by SiXoS — CHANGES_REQUESTED** [review]

File: src/objects/GameObject.cpp [hunk 11]
```diff
@@ -68,18 +46,6 @@ GameObject::GameObject(std::shared_ptr<Mesh> mesh, std::shared_ptr<Mesh> collide
...
-    this->parent = parent;
-    m_modelMatrix = chag::make_identity<chag::float4x4>();
-    initGameObject(mesh, mesh);
-}
-
-GameObject::GameObject(std::shared_ptr<Mesh> mesh, std::shared_ptr<Mesh> colliderMesh, GameObject* parent) {
-    this->parent = parent;
-    m_modelMatrix = chag::make_identity<chag::float4x4>();
-    initGameObject(mesh, colliderMesh);
-}
```
Comment (Bubbad) [inline src/objects/GameObject.cpp:81]: Why were these constructors removed?

File: src/objects/GameObject.cpp [hunk 12]
```diff
@@ -332,7 +297,12 @@ void GameObject::setRotation(chag::Quaternion r) {
     hasRotation = changed = true;
 }
 
-void GameObject::addChild(GameObject* child) {
+void GameObject::addChild(std::shared_ptr<GameObject> child) {
+    if (child->parent) {
+        throw std::invalid_argument("Given child already has a parent");
+    } else {
+        child->parent = shared_from_this();
+    }
```
Comment (Bubbad) [inline src/objects/GameObject.cpp:305]: Perhaps this method should be renamed? It does more than adding children. Something like connectChildAndSetThisAsParent? Long, but at least it's very descriptive

**Review by Bubbad — CHANGES_REQUESTED** [review]

**Comment by DunderRoffe** [comment]
Will have to rework this...
Somehow movement is not stopped by collision...

**Pull request closed without merging**